File: Python/boj/Gold/1941.py
import sys
from itertools import combinations

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    seat = [list(input()) for _ in range(5)]
    seatIndex = [[i, j] for i in range(5) for j in range(5)]
    combi = [list(i) for i in combinations(seatIndex, 7)]
    result = 0

    def isMakeTeam(students):
        count = sum(1 for [x, y] in students if seat[x][y] == "Y")

        return count < 4

    def isConnected(students):
        queue = [students[0]]
        visited = [[False for _ in range(5)] for _ in range(5)]
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]
        count = 1

        visited[students[0][0]][students[0][1]] = True

        while len(queue):
            x, y = queue.pop(0)

            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]

                if nx >= 0 and nx < 5 and ny >= 0 and ny < 5:
                    if not visited[nx][ny] and [nx, ny] in students:
                        queue.append([nx, ny])
                        visited[nx][ny] = True
                        count += 1

        return count == 7

    for c in combi:
        if isMakeTeam(c) and isConnected(c):
            result += 1

    print(result)

File: JavaScript/boj/Gold/1941.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const seat = [];
const seatIndex = [];
const combi = [];
let result = 0;

for (let i = 0; i < 5; i++) {
  for (let j = 0; j < 5; j++) seatIndex.push([i, j]);
  seat.push(input[i].split(''));
}

const isMakeTeam = (students) => {
  const count = students.filter(([x, y]) => seat[x][y] === 'Y').length;

  return count > 3 ? false : true;
};

const isConnected = (students) => {
  const queue = [students[0]];
  const visited = Array.from({ length: 5 }, () => new Array(5).fill(false));
  const dx = [-1, 1, 0, 0];
  const dy = [0, 0, -1, 1];
  let count = 1;

  visited[students[0][0]][students[0][1]] = true;

  while (queue.length) {
    const [x, y] = queue.shift();

    for (let i = 0; i < 4; i++) {
      const nx = x + dx[i];
      const ny = y + dy[i];

      if (nx >= 0 && nx < 5 && ny >= 0 && ny < 5) {
        if (
          !visited[nx][ny] &&
          students.some(([sx, sy]) => sx === nx && sy === ny)
        ) {
          queue.push([nx, ny]);
          visited[nx][ny] = true;
          count++;
        }
      }
    }
  }

  return count === 7 ? true : false;
};

const dfs = (depth, start) => {
  if (depth === 7) {
    if (isMakeTeam(combi) && isConnected(combi)) result++;

    return;
  }

  for (let i = start; i < 25; i++) {
    combi.push(seatIndex[i]);
    dfs(depth + 1, i + 1);
    combi.pop();
  }
};

dfs(0, 0);

console.log(result);
